### PR TITLE
Update api and structure to support json schema

### PR DESCRIPTION
**Pull request by norbye**
__tl;dr: these PRs enable admins to customize info text, what questions are asked and how many questions are asked (with no need of Webkom supervision - hopefully)__

Chunky changes, so I tried to spread it out a little to see if it makes it simpler to review. In case that is not the case, the complete code is available in branch https://github.com/webkom/admissions/tree/json-schema-complete. It might be necessary to run the code from this branch if you want to test it locally, as the codesplitting between the three PRs might be faulty.

This PR is the main bulk of functionality changes, and is required to make the other PRs run and make sense.

- API and structure changes (this PR) #1179
- User-view changes #1179 
- Admin-view changes #1180

There is still plenty to be desired, so this is in no way perfect - but this should be a decent MVP with all necessary functionality intact, and enough to merge it into staging and continue working on it.

Testing is also non-existent, but I'll get to that later..

# Changes
Previously, admissions were hardcoded so every group had a `response_label` string - which is the text prompt that applying users see over the textarea. Each application and group application then had a stored `text` string, which was whatever the user wrote into the application.

Now, I have restructured it so that each group contains a `questions` JSON object, and each application and group application contains a `responses` JSON object.

The questions-object consists of what the group admins choose for it to be, configured through the admin panel (PR #1180).  
So far this can either be a text (normal paragraph), textarea (long text input), textinput (short text) or a phoneinput (norwegian phone number). Each question contains an `id`, `name` and `label` that is used to display the question.

The responses-object consists of whatever was inserted into the application, in a pair of `id` and `value` - where the id matches the id from the corresponding question.

### Example of the JSON bodies
Questions;
```json
[{
    "type": "textarea",
    "id": "550e8400-e29b-41d4-a716-446655440000",
    "name": "Søknadstekst",
    "label": "Hvorfor har du lyst til å søke denne gruppen? Skriv litt om din motivasjon."
}]
```
Responses;
```json
{"550e8400-e29b-41d4-a716-446655440000": "Jeg har kjempelyst ..."}
```

Resolves ABA-513

### CODE REVIEW

File: admissions/admissions/views.py [hunk 1]
```diff
@@ -97,13 +98,15 @@ def get_queryset(self):
...
-            )
+        admission = Admission.objects.get(slug=admission_slug)
+        for group in admission.admin_groups.all():
+            if Membership.objects.filter(user=user, group=group.pk).exists():
+                return (
+                    super()
+                    .get_queryset()
+                    .filter(admission__slug=admission_slug)
+                    .prefetch_related("group_applications", "group_applications__group")
+                )
```
Comment (norbye) [inline admissions/admissions/views.py:109]: Strictly speaking not relevant to this PR, but it was a bug left behind from earlier that prevented me from being able to test the new functionality.  
Previously is_superuser defined whether a user could view all applications, but now that is defined on an application level through the `admin_groups` property

File: admissions/utils/management/commands/create_admission.py [hunk 2]
```diff
@@ -25,6 +25,7 @@ def handle(self, *args, **options):
 
         admission = Admission.objects.create(
             title=f"Opptak {base_date.year}",
+            slug={base_date.year},
```
Comment (norbye) [inline admissions/utils/management/commands/create_admission.py:28]: Required to make fixture work after I added the `slug` field in an earlier PR

File: frontend/src/index.tsx [hunk 3]
```diff
@@ -35,6 +35,10 @@ const queryClient = new QueryClient({
       refetchOnWindowFocus: process.env.NODE_ENV === "production",
       queryFn: defaultQueryFn,
       staleTime: 60000,
+      networkMode: "always",
+    },
+    mutations: {
+      networkMode: "always",
```
Comment (norbye) [inline frontend/src/index.tsx:41]: Added to enable running the project offline. If these flags are not set, react query will not even try to execute the fetch queries if you have no internet connection.

File: admissions/admissions/serializers.py [hunk 4]
```diff
@@ -205,14 +232,14 @@ class Meta:
...
-    def get_text(self, obj):
-        is_filtered = getattr(obj, "group_applications_filtered", False)
-        if is_filtered:
-            return None
-        return obj.text
+    # def get_text(self, obj):
+    #     is_filtered = getattr(obj, "group_applications_filtered", False)
+    #     if is_filtered:
+    #         return None
+    #     return obj.text
```
Comment (norbye) [inline admissions/admissions/serializers.py:242]: Whopsies... Will delete this before merging

File: admissions/admissions/serializers.py [hunk 5]
```diff
@@ -247,33 +273,61 @@ class Meta:
...
+            raise serializers.ValidationError({"groupResponses": {group_name: error}})
+
+    def _validate_fields(self, questions: dict, responses: dict) -> dict:
+        # Ensure that the received JSON indata matches the requested questions
+        editable_questions = list(filter(lambda x: x["type"] != "text", questions))
+        if len(editable_questions) != len(responses):
+            return "Alle spørsmål ble ikke sendt"
+        for question in editable_questions:
+            print(responses)
+            print(question)
```
Comment (LudvigHz) [inline admissions/admissions/serializers.py:300]: :)

File: admissions/admissions/serializers.py [hunk 6]
```diff
@@ -247,33 +273,61 @@ class Meta:
...
+            response_key = next(
+                (key for key in responses.keys() if key == question["id"]), None
+            )
+            if response_key is None:
+                return {question["id"]: "Obligatorisk felt ble ikke sendt"}
+            response = responses[response_key]
+            if response is None:
+                return {question["id"]: "Obligatorisk felt"}
```
Comment (LudvigHz) [inline admissions/admissions/serializers.py:308]: Hmm, this looks like it could be quite slow? Not a huge deal here though. Not exactly performance critical software :grimacing:

**Review by LudvigHz — APPROVED** [review]
Skimmed through on my phone. Looks good :)

Although this seems like it would be a lot easier to do using pydantic (schema definition, types, validation etc) but this works so I won't force you to do anything. But worth keeping in mind for similar usecases later.

File: admissions/admissions/serializers.py (hunk 6 above)
Comment (norbye) [reply admissions/admissions/serializers.py:308]: Hmm yeah.. Do you have any ideas from the top of your head about how to improve it? Could be nice to improve, but yeah it's not very ciritical and it's run very few times per application

**Comment by norbye** [comment]
> Although this seems like it would be a lot easier to do using pydantic (schema definition, types, validation etc) but this works so I won't force you to do anything. But worth keeping in mind for similar usecases later.

Wow thanks a lot, that looks like a much better option! Will take a look at it

**Comment by norbye** [comment]
Converted to draft to ensure it is not merged before 2023 admissions, will merge some time after that.

**Comment by norbye** [comment]
This was quite annoying to rebase at this point, so I'm closing it off for now - but keeping the branch so I can look into it at a later time, possibly changing the backend to use pydantic(:
